packages/unified-lint-rule/index.d.ts

```diff
@@ -18,17 +18,13 @@ export interface RuleMeta {
 export function lintRule<Tree extends Node = Node, Settings = unknown>(
   name: string | RuleMeta,
   rule: Rule<Tree, Settings>
-): Plugin<
-  | void[]
-  | [Settings | Label | Severity]
-  | [boolean | Label | Severity, Settings],
-  Tree
->
+): Plugin<[(Settings | [Label | Severity, Settings?] | void)?], Tree>
```

Comment:
To avoid getting:
```sh
lib/index.js:89:62 - error TS2345: Argument of type '[void | Options]' is not assignable to parameter of type 'void[] | [Options | Label | Severity] | [boolean | Label | Severity, Options]'.
  Type '[void | Options]' is not assignable to type '[Options | Label | Severity]'.
    Type 'void | Options' is not assignable to type 'Options | Label | Severity'.
      Type 'void' is not assignable to type 'Options | Label | Severity'.
```
calling `lintRule(...)(rawOptions)` [here](https://github.com/remarkjs/remark-validate-links/pull/67/files#diff-92bbac9a308cd5fcf9db165841f2d90ce981baddcb2b1e26cfff170929af3bd1R89).

I think the correct type is `[(Settings | [Label | Severity, Settings?])?]` (no `void`), however the inferred type [here](https://github.com/unifiedjs/unified/blob/6b060c2a229049e1bc0e7ea51920b36efb069f9f/index.d.ts#L135) becomes `[Settings | [Label | Severity, Settings | undefined] | undefined]`, which requires too many elements.

I tried with TS 4.6 and [`--exactOptionalPropertyTypes`](https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes) but couldn't get it to infer the optional elements (`[(Settings | [Label | Severity, Settings?])?]`). I haven't yet investigated further, whether this corresponds to an open TS issue.

I suspect this is the reason that `| void[]` was there before.

Reply:
I didn't find an existing TS issue for the optional parameters, so I opened one: https://github.com/microsoft/TypeScript/pull/48132

Reply:
`Array<void>` indeed tries to match the case where no arguments are given. The reason is that `[]` is used for tuples. In practise, amongst most plugins, I’ve had the most luck using `Array<void>`.

Your changes here miss the `boolean` case and IMO are a harder to read than before.
Am I correct that without these changes, you got the error you mentioned after “To avoid getting:”? Can we perhaps solve that another way?

Reply:
I think `void` needs to be inside the tuple: `[... | void]` vs. `[...] | Array<void>`. Logically `[... | void]` is assignable to `[...] | Array<void>` (because you can't have a heterogeneous 1-tuple) but TS doesn't support it, currently: https://github.com/microsoft/TypeScript/issues/30895

I'm calling `lintRule(...).call(this, rawOptions)`, where `rawOptions` is `Options | void`.

Plugins aren't called with `boolean`? https://github.com/unifiedjs/unified/blob/2323d38e2c74708f5a453ebfea42f80e0454a435/lib/index.js#L127-L133

`processor.use(plugin[, options])` adds the `boolean` type, as it should: https://github.com/unifiedjs/unified/blob/6b060c2a229049e1bc0e7ea51920b36efb069f9f/index.d.ts#L136

`?` allows an explicit `undefined` if [`--exactOptionalPropertyTypes`](https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes) is *off*, otherwise not. TS currently allows an explicit `undefined` *argument* regardless, however, probably because it uses one signature for callers and for the function body? Dropping `| undefined` in the function body would be wrong if an optional argument were omitted. The docs agree: https://www.typescriptlang.org/docs/handbook/2/functions.html#optional-parameters

Pondering out loud: Maybe TS could drop `| undefined` from function overloads, since those are only for callers, currently? It doesn't, maybe in case that changes?

Here are playground links [with](https://www.typescriptlang.org/dev/bug-workbench/?exactOptionalPropertyTypes=true#code/PTAEAEFMA8EMGMAuB5ADoglgewHawDYAKATlqpMYgJ4AqV5AzgFBP6SKhma4MBcoAbQBEXbHnxCA-AF0A3E1E9QAXkEj0YgkLmgQoQCjkgeD-QAOSygKpYgwA0oWAwvRySSABMFGpaoE69R0ABRNgBbSBwODEdFcU9uHEcfAFccN0gAMwwcdz8wQBlyUBS0zOy3UCicAHIOBwYMAHM8ACM2UERzdXitFnSUpDFQVHwk+qyACgA6KZi+NRitGQBKfgA3LAw3WV0wGkgGDiwVinwsWDK6xthEJOJIJl6cftxB4dGcSemvBP5heYkl0AAbwAvkwhiNxp1NBJFlt-MYzBZiFZbPZHDAXIh3GDXuNYdsDMZCLBiLAwljiOVol8CDiIe8ihksu58XpGSV3FT7PgTgB3dx2W71EluNgMBhAA) and [without](https://www.typescriptlang.org/dev/bug-workbench/?exactOptionalPropertyTypes=true#code/DYUwLgBA9gDmCWUB2BnAXBA2gIlg5AhsNgPwC6A3AFB6KoQC8WucdR2lEA9FxICjkgeD+IAOSgQQAJwlQJKADQQCKcQA8YIAMZgQAExqtkyppk49+QgKKgAtiCSR4y2oWD789YwFckOkADN4JF1TXkERMUlpWQh4PwgAWniQFQItAHkDJCIABWl1CTAATwAVQvVlR2g-PyoqP28tOggYYE8Ac0CACgA6XudUDBx+9nIASgwANyh4HQpuXmKQFEgoCclgKAIdCBR4NqywTwkQOob3ZtaOpB6+zPRmYeIxiABvAF8qFvaulnd2UbmZjConEUhk8kUymS6i0uk+ly6APm5gg2QIEgItm0EhiTkyRHh32u3l8ASCOiRZhJ-kCulximAGwA7roFMc2uidKAUCggA) `--exactOptionalPropertyTypes`:
```ts repro
// @exactOptionalPropertyTypes

let options: ["optional"?];
options = ["optional"]; // ✔️ No errors, as expected
options = []; // ✔️ Element is optional
options = [undefined]; // ❌ undefined isn't assignable to "optional"

function plugin(...options: ["optional"?]): void; // Test overload signature
function plugin(...options: ["optional"?]) {}
plugin("optional"); // ✔️ No errors, as expected
plugin(); // ✔️ Parameter is optional
plugin(undefined); // undefined is allowed, regardless

Reply:
> I think void needs to be inside the tuple:

I’ve had troubles with trying to type plugins that way, and calling them without arguments.
Only `Array<void>` seems to allow all cases, including no parameter.

> I'm calling `lintRule(...).call(this, rawOptions)`, where `rawOptions` is `Options | void`.

Where? This PR does not include that code. Perhaps you can, on the using side, make sure that `void` is removed from that value, e.g.:

```
if (rawOptions) {
  // ...
}
```

> Plugins aren't called with `boolean`?

You are correct. Though the intent is that it’s about an *array in the tuple*:

```
.use(someRule, true) // this one is not needed, indeed
.use(someRule, [true]) // this one *is* needed, though
.use(someRule, [true, 72]) // same
```

> `?` allows an explicit `undefined` if `--exactOptionalPropertyTypes` is off

If possible, I’d prefer to code explicitly in a way that is not affected by such options.
I’m not sure about others but I find it hard to write code in different projects if I have to check *what* it actually means by looking at the `tsconfig`? And I don’t expect other contributors to be aware of such small but important details.
That’s why the code is currently explicitly using fixed-size tuples. And when those aren’t possible (0-size tuple), an `Array<void>`.

> Pondering out loud: Maybe TS could drop `| undefined` from function overloads

I don’t know enough about TS to decide what the right decision for them is!

Reply:
> If possible, I’d prefer to code explicitly in a way that is not affected by such options.
> I’m not sure about others but I find it hard to write code in different projects if I have to check _what_ it actually means by looking at the `tsconfig`? And I don’t expect other contributors to be aware of such small but important details.

:+1: That's smart, how about explicitly adding `undefined` to the optional elements? i.e. `[("optional" | undefined)?]`
That's the same with and without `--exactOptionalPropertyTypes`.

> I’ve had troubles with trying to type plugins that way, and calling them without arguments.
> Only `Array<void>` seems to allow all cases, including no parameter.

I see what you mean when they aren't optional. When they are optional, it isn't needed any more.

> ```
> .use(someRule, [true]) // this one *is* needed, though
> .use(someRule, [true, 72]) // same
> ```

Are these cases supported?

> Where? This PR does not include that code. Perhaps you can, on the using side, make sure that `void` is removed from that value, e.g.:
> 
> ```
> if (rawOptions) {
>   // ...
> }
> ```

Here's the using side: https://github.com/remarkjs/remark-validate-links/pull/67/files#diff-92bbac9a308cd5fcf9db165841f2d90ce981baddcb2b1e26cfff170929af3bd1R93

There, as here, if we can avoid introducing `void` in the first place, that's the simplest and least error-prone way to remove `void`?